In [51]:
def initialize(r):
    arr =[[0,0,0] for i in range(3)]
    arr[0][2]=10
    arr[0][0]=r
    array=iteration_function(arr,r)
    for row in array:
        print(row)
    policy =optain_policy(array,r)
    for row in policy:
        print(row)
    
   
          

    

Calculate Reward

In [36]:
def calculate_reward(x,y,r):
   if(x==0 and y==0):
        return r
   elif(x==0 and y==2):
        return 10
   else:
        return -1
         
  

Calculate Transition function

In [34]:

def calc_transition(x1,y1,a,x2,y2):
    # new state is up
    if(x2==x1-1 and y2==y1):
        # action is up
        if(a==0):
            return round(0.8,5)
        # action is right
        elif(a==2):
            return round(0.1,5)
        # action is left
        elif(a==3):
            return round(0.1,5)
    # new state is down    
    elif(x2==x1+1 and y2==y1):
        # action is down
        if(a==1):
            return round(0.8,5)
        # action is right
        elif(a==2):
            return round(0.1,5)
        # action is left
        elif(a==3):
            return round(0.1,5)
    # new state is right    
    elif(x2==x1 and y2==y1+1):
        # action is right
        if(a==2):
            return round(0.8,5)
        # action is up
        elif(a==0):
            return round(0.1,5)
        # action is down
        elif(a==1):
            return round(0.1,5)
    # new state is left
    elif(x2==x1 and y2==y1-1):
        # action is left
        if(a==3):
            return round(0.8,5)
        # action is up
        elif(a==0):
            return round(0.1,5)
        # action is down
        elif(a==1):
            return round(0.1,5)
            

             
       
                






In [29]:
def max_absolute_difference(array1, array2):
    # Ensure that the arrays have the same shape
    if len(array1) != len(array2) or len(array1[0]) != len(array2[0]):
        raise ValueError("Arrays must have the same shape")

    max_diff = 0

    for i in range(len(array1)):
        for j in range(len(array1[0])):
            diff = abs(array1[i][j] - array2[i][j])
            max_diff = max(max_diff, diff)

    return max_diff

In [56]:



actions = [0, 1, 2, 3]
movements = {0: (-1, 0), 1: (1, 0), 2: (0, 1), 3: (0, -1)}

def iteration_function(arr, r):
    # print(arr)
    delta = 0
    # i=10
    while True:
    # for i in range(1):
        new_arr =[[0,0,0] for i in range(3)]
        new_arr[0][2]=10
        new_arr[0][0]=r
        for i in range(3):
            for j in range(3):
                if (i == 0 and j == 2) or (i == 0 and j == 0):
                    continue
                else:
                    values = []
                    for action in actions:
                        ci, cj = i +  movements[action][0], j + movements[action][1]
                        r1, r2=get_right_angels(action)
                        ci1,cj1=i+ movements[r1][0], j + movements[r1][1]
                        ci2,cj2= i +  movements[r2][0], j+ movements[r2][1]
                        result=Calc_Belman_eq(i,j,ci,cj,ci1,cj1,ci2,cj2,action,arr,r)
                        values.append(result)
                    new_arr[i][j] = max(values)
        
        delta =max_absolute_difference(arr,new_arr)
        if delta<0.0000001:
            return arr
        arr=new_arr

        # for row in arr:
        #     print (row)
        # print()


    
def get_right_angels(action):
    if action == 0:
        return 2,3
    if action == 1:
        return 2,3
    if action == 2:
        return 0,1
    if action == 3:
        return 0,1
    


initialize(-3)


[-3, 8.46194, 10]
[4.78304, 7.11321, 8.46194]
[4.50885, 5.79413, 6.96502]
['T', 'R', 'T']
['R', 'R', 'U']
['R', 'R', 'U']


In [42]:
def Calc_Belman_eq(i,j,ci,cj,ci1,cj1,ci2,cj2,action,arr,r):
    if(0 <= ci < 3 and 0 <= cj < 3):
     result=round(calc_transition(i,j,action,ci,cj) * ( round(calculate_reward(i,j,r),5) + round(0.99 * arr[ci][ cj],5)),5)
     result=round(result,5)
    else:
     result=round(calc_transition(i,j,action,ci,cj) * ( round(calculate_reward(i,j,r),5) + round(0.99 * arr[i][j],5)),5)
     result=round(result,5) 
    if 0 <= ci1 < 3 and 0 <= cj1 < 3:
     result=round(result,5)
     result+=round(calc_transition(i,j,action,ci1,cj1) * ( calculate_reward(i,j,r) + round(0.99 * arr[ci1][cj1],5)),5)
    else:
     result=round(result,5)
     result+=round(calc_transition(i,j,action,ci1,cj1) * ( calculate_reward(i,j,r) + round(0.99 * arr[i][j],5)),5)    
    if 0 <= ci2 < 3 and 0 <= cj2 < 3:
     result=round(result,5)
     result+=round(calc_transition(i,j,action,ci2,cj2) * ( calculate_reward(i,j,r) + round(0.99 * arr[ci2][cj2],5)),5)
    else:
     result=round(result,5)
     result+=round(calc_transition(i,j,action,ci2,cj2) * ( calculate_reward(i,j,r) + round(0.99 * arr[i][j],5)),5)
                            
    result=round(result,5)
    return result

In [49]:
def optain_policy(arr,r):
        policy =[["","",""] for i in range(3)]
        for i in range(3):
            for j in range(3):
                if (i == 0 and j == 2) or (i == 0 and j == 0):
                    policy[i][j] = "T"
                    continue
                else:
                    values = []
                    max_value=-100000000
                    max_action=0
                    for action in actions:
                        ci, cj = i +  movements[action][0], j + movements[action][1]
                        r1, r2=get_right_angels(action)
                        ci1,cj1=i+ movements[r1][0], j + movements[r1][1]
                        ci2,cj2= i +  movements[r2][0], j+ movements[r2][1]
                        result=Calc_Belman_eq(i,j,ci,cj,ci1,cj1,ci2,cj2,action,arr,r)
                        if(result>max_value):
                            max_value=result
                            max_action=action
                    if(max_action==0):
                        policy[i][j]="U"
                    elif(max_action==1):
                        policy[i][j]="D"
                    elif(max_action==2):
                        policy[i][j]="R"
                    elif(max_action==3):
                        policy[i][j]="L"
        return policy

-------------------------------------------------------Report----------------------------------------------------------------------------
1) r=100:
V*(s) = [100, 97.20353, 10]
        [97.20353, 94.75128, 88.20426]
        [94.48183, 92.35293, 89.7622]
        
optimal policy for each state = ['T', 'L', 'T']
                                ['U', 'L', 'D']
                                ['U', 'L', 'L']  

2) r=3:
V*(s) = [3, 8.46194, 10]
        [5.38357, 7.11321, 8.46194]
        [4.57483, 5.79413, 6.96502]
        
optimal policy for each state = ['T', 'R', 'T']
                                ['R', 'R', 'U']
                                ['R', 'R', 'U']

3) r=0:
V*(s) = [0, 8.46194, 10]
        [5.0833, 7.11321, 8.46194]
        [4.54184, 5.79413, 6.96502]
        
optimal policy for each state = ['T', 'R', 'T']
                                ['R', 'R', 'U']
                                ['R', 'R', 'U']

4) r=-3:
V*(s) = [-3, 8.46194, 10]
        [4.78304, 7.11321, 8.46194]
        [4.50885, 5.79413, 6.96502]
        
optimal policy for each state = ['T', 'R', 'T']
                                ['R', 'R', 'U']
                                ['R', 'R', 'U']                               